qcfractal/interface/collections/dataset_view.py

```diff
@@ -0,0 +1,259 @@
+        units = {}
+        with self._read_file() as f:
+            ret = pd.DataFrame(index=f["entry"][:])
+
+            for query in queries:
+                dataset_name = "value/" if query["native"] else "contributed_value/"
+                dataset_name += self._normalize_hdf5_name(query["name"])
+                driver = query["driver"]
+
+                dataset = f[dataset_name]
+                if not h5py.check_dtype(vlen=dataset.dtype):
+                    data = list(dataset[:])
+                else:
+                    nentries = dataset.shape[0]
+                    if driver.lower() == "gradient":
+                        data = [np.reshape(dataset[i], (-1, 3)) for i in range(nentries)]
+                    elif driver.lower() == "hessian":
+                        data = []
+                        for i in range(nentries):
+                            n2 = len(dataset[i])
+                            n = int(round(np.sqrt(n2)))
+                            data.append(np.reshape(dataset[i], (n, n)))
+                    else:
+                        warnings.warn(
+                            f"Variable length data type not understood, returning flat array "
+                            f"(driver = {driver}).", RuntimeWarning)
+                        data = list(dataset[:])
+                column_name = query["name"]
+                column_units = self._deserialize_field(dataset.attrs["units"])
+                ret[column_name] = data
+                units[column_name] = column_units
+
+        return ret, units
+
+    def write(self, ds: Dataset):
+        # For data checksums
+        dataset_kwargs = {"chunks": True, "fletcher32": True}
+
+        n_records = len(ds.data.records)
+        default_shape = (n_records, )
+
+        if h5py.__version__ >= distutils.version.StrictVersion("2.10.0"):
+            vlen_double_t = h5py.vlen_dtype(np.dtype("float64"))
+            utf8_t = h5py.string_dtype(encoding="utf-8")
+            vlen_utf8_t = h5py.vlen_dtype(utf8_t)
+        else:
+            vlen_double_t = h5py.special_dtype(vlen=np.dtype("float64"))
+            utf8_t = h5py.special_dtype(vlen=str)
+            vlen_utf8_t = h5py.special_dtype(vlen=utf8_t)
+
+        driver_dataspec = {
+            "energy": {
+                "dtype": np.dtype("float64"),
+                "shape": default_shape
+            },
+            "gradient": {
+                "dtype": vlen_double_t,
+                "shape": default_shape
+            },
+            "hessian": {
+                "dtype": vlen_double_t,
+                "shape": default_shape
+            },
+            "dipole": {
+                "dtype": np.dtype("float64"),
+                "shape": (n_records, 3)
+            }
+        }
+
+        def _write_dataset(dataset, column, entry_dset):
+            assert column.shape[1] == 1
+            for i, name in enumerate(entry_dset):
```

Comment:
Looping through this data seems slow, is there any way to do this in a vectorized way?

Reply:
As far as I could tell, the only data which can be written in a vectorized way are fixed length. Unfortunately, most of our large objects (geometries, gradients, hessians, orbitals) will be variable length. For now, I'm not worrying too much about write performance since writes should be rare.

Reply:
Sounds good!